File: wc111/case2/case2.go
package main

func diStringMatch(S string) []int {
	var n = len(S) + 1
	var path = []int{0}
	var left = []int{}
	for i := 1; i < n; i++ {
		left = append(left, i)
	}

	if n == 1 {
		return []int{}
	}

	var found bool
	var dfs func(int, []int, []int) []int

	dfs = func(ll int, path []int, left []int) []int {
		if len(left) == 0 {

		}
		for !found {
			var letter = S[ll]
			switch letter {
			case 'I':
				for i, v := range left {
					if v > path[len(path)-1] {
						dfs(ll+1, append(path, v), append(left[:i], left[i+1:]...))
					}
				}
			default:
				for i, v := range left {
					if v < path[len(path)-1] {
						dfs(ll+1, append(path, v), append(left[:i], left[i+1:]...))
					}
				}
			}

		}
		return path
	}
}

func main() {}

File: wc111/case2.py
class Solution:
    def diStringMatch(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        N = len(S)

        def dfs(start, level, used, path):
            # print(start, level, used, path)
            if level == N:
                return path
            ss = S[level]
            for i in range(N+1):
                if ss == 'I' and start < i and i not in used:
                    newused = {i for i in used}
                    newpath = [i for i in path]
                    newused.add(i)
                    newpath.append(i)
                    v = dfs(i, level+1, newused, newpath)
                    if v:
                        return v
                elif ss == 'D' and start > i and i not in used:
                    newused = {i for i in used}
                    newpath = [i for i in path]
                    newused.add(i)
                    newpath.append(i)
                    v = dfs(i, level+1, newused, newpath)
                    if v:
                        return v

        for start in range(N+1):
            v = dfs(start, 0, set([start]), [start])
            if v:
                return v

print(Solution().diStringMatch("DIDI"))
print(Solution().diStringMatch("IDID"))


